pkg/storegateway/sharding_strategy_test.go

```diff
@@ -499,3 +500,345 @@ type shardingLimitsMock struct {
+	removeInstancesFromRing := func(d *ring.Desc, instances []instanceDef) {
+		for _, inst := range instances {
+			d.RemoveIngester(inst.id)
+		}
+	}
+
+	// Helper to get instance IDs from definitions.
+	getInstanceIDs := func(instances ...[]instanceDef) []string {
+		var ids []string
+		for _, instList := range instances {
+			for _, inst := range instList {
+				ids = append(ids, inst.id)
+			}
+		}
+		return ids
+	}
+
+	// ============================================================================
+	// Step 1: Initial state - 3 zones, RF=3
+	// ============================================================================
+	t.Log("Step 1: Setting up initial ring with 3 zones, RF=3")
+
+	require.NoError(t, store.CAS(ctx, "test", func(interface{}) (interface{}, bool, error) {
+		d := ring.NewDesc()
+		addInstancesToRing(d, zoneAInstances)
+		addInstancesToRing(d, zoneBInstances)
+		addInstancesToRing(d, zoneCInstances)
+		return d, true, nil
+	}))
+
+	ringCfg := ring.Config{
+		ReplicationFactor:    3,
+		HeartbeatTimeout:     time.Minute,
+		SubringCacheDisabled: true,
+		ZoneAwarenessEnabled: true,
+	}
+
+	r, err := ring.NewWithStoreClientAndStrategy(ringCfg, "test", "test", store, ring.NewIgnoreUnhealthyInstancesReplicationStrategy(), nil, log.NewNopLogger())
+	require.NoError(t, err)
+	require.NoError(t, services.StartAndAwaitRunning(ctx, r))
+
+	// Track the current ring for cleanup. We'll update this as we create new rings.
+	currentRing := r
+	t.Cleanup(func() { _ = services.StopAndAwaitTerminated(ctx, currentRing) })
+
+	// Wait until the ring client has synced.
+	require.NoError(t, ring.WaitInstanceState(ctx, r, zoneAInstances[0].id, ring.ACTIVE))
+
+	// Capture baseline state for all instances.
+	allInitialInstances := append(append(zoneAInstances, zoneBInstances...), zoneCInstances...)
+	baselineState := captureState(t, r, allInitialInstances, 3, limits)
+
+	t.Logf("Baseline captured: %d instances", len(baselineState))
+
+	// ============================================================================
+	// Step 2: Increase RF to 4 (same ring state, no instance changes)
+	// ============================================================================
+	t.Log("Step 2: Changing RF from 3 to 4 (no instance changes)")
+
+	// Stop the old ring and create a new one with RF=4.
+	require.NoError(t, services.StopAndAwaitTerminated(ctx, r))
+
+	ringCfg.ReplicationFactor = 4
+	r, err = ring.NewWithStoreClientAndStrategy(ringCfg, "test", "test", store, ring.NewIgnoreUnhealthyInstancesReplicationStrategy(), nil, log.NewNopLogger())
+	require.NoError(t, err)
+	require.NoError(t, services.StartAndAwaitRunning(ctx, r))
+	currentRing = r // Update for cleanup
+
+	// Wait for ring to sync.
+	require.NoError(t, ring.WaitInstanceState(ctx, r, zoneAInstances[0].id, ring.ACTIVE))
+
+	// Verify all instances have the same blocks as before.
+	stateAfterRFChange := captureState(t, r, allInitialInstances, 4, limits)
```

Comment:
[nit] I think it's safer:

```suggestion
	stateAfterRFChange := captureState(t, r, allInitialInstances, ringCfg.ReplicationFactor, limits)
```